crates/runtime/src/instance/allocator/memfd.rs

```diff
@@ -0,0 +1,258 @@
+        // no image but the same initial size, so the mappings are
+        // already correct; there is no need to mmap anything. Given
+        // that we asserted not-dirty above, any dirty pages will have
+        // already been thrown away by madvise() during the previous
+        // termination.  The `clear_and_remain_ready()` path also
+        // mprotects memory above the initial heap size back to
+        // PROT_NONE, so we don't need to do that here.
+        if (self.image.is_none()
+            && maybe_image.is_none()
+            && self.initial_size == initial_size_bytes)
+            || (self.image.is_some()
+                && maybe_image.is_some()
+                && self.image.as_ref().unwrap().fd.as_file().as_raw_fd()
+                    == maybe_image.as_ref().unwrap().fd.as_file().as_raw_fd())
+        {
+            self.dirty = true;
+            return Ok(());
+        }
+
+        // Otherwise, we need to redo (i) the anonymous-mmap backing
+        // for the whole slot, (ii) the initial-heap-image mapping if
+        // present, and (iii) the mprotect(PROT_NONE) above the
+        // initial heap size.
+
+        // Security/audit note: we map all of these MAP_PRIVATE, so
+        // all instance data is local to the mapping, not propagated
+        // to the backing fd. We throw away this CoW overlay with
+        // madvise() below, from base up to static_size (which is the
+        // whole slot) when terminating the instance.
+
+        // Anonymous mapping behind the initial heap size: this gives
+        // zeroes for any "holes" in the initial heap image. Anonymous
+        // mmap memory is faster to fault in than a CoW of a file,
+        // even a file with zero holes, because the kernel's CoW path
+        // unconditionally copies *something* (even if just a page of
+        // zeroes). Anonymous zero pages are fast: the kernel
+        // pre-zeroes them, and even if it runs out of those, a memset
+        // is half as expensive as a memcpy (only writes, no reads).
+        if initial_size_bytes > 0 {
+            unsafe {
+                let ptr = rustix::io::mmap_anonymous(
+                    self.base as *mut c_void,
+                    self.static_size,
+                    rustix::io::ProtFlags::READ | rustix::io::ProtFlags::WRITE,
+                    rustix::io::MapFlags::PRIVATE | rustix::io::MapFlags::FIXED,
+                )
+                .map_err(|e| InstantiationError::Resource(e.into()))?;
+                assert_eq!(ptr as usize, self.base);
+            }
+        }
+
+        // The initial memory image, if given. If not, we just get a
+        // memory filled with zeroes.
+        if let Some(image) = maybe_image {
+            if image.len > 0 {
+                let image = image.clone();
+
+                unsafe {
+                    let fd = rustix::fd::BorrowedFd::borrow_raw_fd(image.fd.as_file().as_raw_fd());
```

Comment:
as a minor nit the traits in rustix seem like you could probably pass `image.fd.as_file()` below as the `fd` argument rather than creating a new `BorrowedFd` structure (although I haven't used rustix, just a hunch)

Reply:
Good call, thanks!